File: services/routing.ts
import {
  Building,
  RouteData,
  RouteSegment,
  AccessibilityProfile,
  Coordinates,
} from '@/types';
import { BUILDINGS, PATHS, SAMPLE_EVENTS } from '@/constants/campus';
import {
  getPathAttributes,
  CAMPUS_PULSE,
  PULSE_AFFECTED_BUILDINGS,
} from '@/constants/campusFeatures';
import { isBadWeather } from '@/services/weather';

export type RouteType = 'fastest' | 'accessible' | 'scenic' | 'quiet' | 'weather_shielded';

interface GraphEdge {
  to: string;
  from: string;
  distance: number;
  walkTime: number;
  isAccessible: boolean;
  hasStairs: boolean;
  pathId: string;
  noiseLevel: number;
  isCovered: boolean;
  isIndoor: boolean;
}

const BUILDING_BY_ID = new Map<string, Building>(
  BUILDINGS.map((b) => [b.building_id, b])
);

function pathId(from: string, to: string): string {
  return `${from}-${to}`;
}

function buildGraph(): Map<string, GraphEdge[]> {
  const graph = new Map<string, GraphEdge[]>();

  const addEdge = (from: string, to: string, base: Omit<GraphEdge, 'to' | 'from' | 'pathId'>) => {
    if (!graph.has(from)) graph.set(from, []);
    graph.get(from)!.push({ ...base, from, to, pathId: pathId(from, to) });
  };

  for (const p of PATHS) {
    const attr = getPathAttributes(p.from, p.to);
    const shared = {
      distance: p.distance,
      walkTime: p.walk_time,
      isAccessible: p.is_accessible,
      hasStairs: p.has_stairs,
      noiseLevel: attr.noise_level,
      isCovered: attr.is_covered,
      isIndoor: attr.is_indoor,
    };
    addEdge(p.from, p.to, shared);
    addEdge(p.to, p.from, shared);
  }

  return graph;
}

const GRAPH = buildGraph();

const PULSE_BY_BUILDING = new Map(CAMPUS_PULSE.map((p) => [p.building_id, p]));

function getBlockedPathIds(): Set<string> {
  const blocked = new Set<string>();
  for (const event of SAMPLE_EVENTS) {
    if (!event.is_active || !event.affects_navigation || !event.blocked_paths) continue;
    for (const bp of event.blocked_paths) {
      blocked.add(bp);
      const [a, b] = bp.split('-');
      if (a && b) {
        blocked.add(pathId(a, b));
        blocked.add(pathId(b, a));
      }
    }
  }
  return blocked;
}

function needsStepFree(profile: AccessibilityProfile): boolean {
  return profile.wheelchair || profile.elevator_required || profile.avoid_stairs;
}

function needsQuiet(profile: AccessibilityProfile, routeType: RouteType): boolean {
  return routeType === 'quiet' || profile.sensory_friendly;
}

function pulsePenalty(edge: GraphEdge): number {
  let penalty = 0;
  for (const node of [edge.from, edge.to]) {
    if (!PULSE_AFFECTED_BUILDINGS.has(node)) continue;
    const pulse = PULSE_BY_BUILDING.get(node);
    if (!pulse) continue;
    if (pulse.status === 'busy') penalty += (pulse.congestion_penalty_minutes ?? 3) * 25;
    else if (pulse.status === 'full') penalty += 200;
    else if (pulse.status === 'moderate') penalty += 40;
  }
  return penalty;
}

function edgeCost(
  edge: GraphEdge,
  routeType: RouteType,
  profile: AccessibilityProfile,
  blocked: Set<string>,
  badWeather: boolean
): number {
  if (blocked.has(edge.pathId)) return Infinity;

  const mustBeStepFree =
    routeType === 'accessible' || routeType === 'weather_shielded' || needsStepFree(profile);

  if (mustBeStepFree && (edge.hasStairs || !edge.isAccessible)) {
    return Infinity;
  }

  let cost = edge.distance + pulsePenalty(edge);

  if (routeType === 'scenic') {
    if (edge.hasStairs) cost += 250;
    if (!edge.isAccessible) cost += 150;
  }

  if (needsQuiet(profile, routeType)) {
    cost += edge.noiseLevel * 35;
    if (edge.noiseLevel >= 7) cost += 120;
  }

  if (routeType === 'weather_shielded' || (badWeather && routeType !== 'fastest')) {
    if (!edge.isCovered && !edge.isIndoor) cost += 180;
    if (edge.isIndoor) cost -= 20;
    if (edge.isCovered) cost -= 10;
  }

  return Math.max(cost, edge.distance * 0.5);
}

function dijkstra(
  fromId: string,
  toId: string,
  costFn: (edge: GraphEdge) => number
): { path: string[]; edges: GraphEdge[] } | null {
  if (fromId === toId) return { path: [fromId], edges: [] };
  if (!GRAPH.has(fromId) || !BUILDING_BY_ID.has(toId)) return null;

  const dist = new Map<string, number>();
  const prev = new Map<string, { node: string; edge: GraphEdge }>();
  const visited = new Set<string>();
  dist.set(fromId, 0);

  while (true) {
    let current: string | null = null;
    let currentDist = Infinity;
    for (const [node, d] of dist) {
      if (!visited.has(node) && d < currentDist) {
        current = node;
        currentDist = d;
      }
    }
    if (current === null || current === toId) break;
    visited.add(current);

    for (const edge of GRAPH.get(current) ?? []) {
      if (visited.has(edge.to)) continue;
      const stepCost = costFn(edge);
      if (!isFinite(stepCost)) continue;
      const next = currentDist + stepCost;
      if (next < (dist.get(edge.to) ?? Infinity)) {
        dist.set(edge.to, next);
        prev.set(edge.to, { node: current, edge });
      }
    }
  }

  if (!dist.has(toId) || !isFinite(dist.get(toId)!)) return null;

  const path: string[] = [];
  const edges: GraphEdge[] = [];
  let node: string | undefined = toId;
  while (node && node !== fromId) {
    path.unshift(node);
    const p = prev.get(node);
    if (!p) return null;
    edges.unshift(p.edge);
    node = p.node;
  }
  path.unshift(fromId);
  return { path, edges };
}

function paceMultiplier(profile: AccessibilityProfile): number {
  if (profile.slow_walker) return 1.4;
  if (profile.wheelchair) return 1.15;
  return 1;
}

function buildSegments(edges: GraphEdge[]): RouteSegment[] {
  return edges.map((e) => ({
    from: e.from,
    to: e.to,
    distance_meters: e.distance,
    walk_time_minutes: e.walkTime,
    path_type: e.hasStairs ? 'footpath' : e.isIndoor ? 'indoor' : e.isCovered ? 'covered' : 'main_road',
    is_accessible: e.isAccessible && !e.hasStairs,
    notes: e.hasStairs
      ? 'Contains stairs'
      : e.noiseLevel >= 7
      ? 'Noisy area — sensory route avoids this when possible'
      : undefined,
  }));
}

function buildSteps(path: string[], routeType: RouteType): string[] {
  if (path.length === 0) return [];
  const startName = BUILDING_BY_ID.get(path[0])?.name ?? path[0];
  const steps: string[] = [`Start at ${startName}`];

  for (let i = 1; i < path.length; i++) {
    const building = BUILDING_BY_ID.get(path[i]);
    const name = building?.short_name ?? path[i];
    const isLast = i === path.length - 1;
    if (isLast) {
      const full = building?.name ?? name;
      if (routeType === 'accessible' || routeType === 'weather_shielded') {
        steps.push(`Arrive at ${full} via the accessible entrance`);
      } else {
        steps.push(`Arrive at ${full}`);
      }
    } else {
      steps.push(`Continue past ${name}`);
    }
  }
  return steps;
}

function buildWaypoints(path: string[]): Coordinates[] {
  return path
    .map((id) => BUILDING_BY_ID.get(id)?.coordinates)
    .filter((c): c is Coordinates => Boolean(c));
}

function collectPulseWarnings(path: string[]): string[] {
  const warnings: string[] = [];
  for (const node of path) {
    const pulse = PULSE_BY_BUILDING.get(node);
    if (pulse && (pulse.status === 'busy' || pulse.status === 'full') && pulse.detail) {
      warnings.push(pulse.detail);
    }
  }
  return [...new Set(warnings)];
}

function collectEventWarnings(): string[] {
  return SAMPLE_EVENTS.filter(
    (e) => e.is_active && e.affects_navigation && e.alternate_route_note
  ).map((e) => e.alternate_route_note as string);
}

function notesForRoute(routeType: RouteType, profile: AccessibilityProfile, badWeather: boolean): string | undefined {
  if (routeType === 'quiet' || profile.sensory_friendly) {
    return 'Sensory-quiet route. Avoids noisy cafeterias, crowded plazas, and high-traffic zones.';
  }
  if (routeType === 'weather_shielded' || (badWeather && routeType === 'accessible')) {
    return 'Weather-shielded route. Maximizes covered walkways and indoor corridors.';
  }
  if (routeType === 'accessible' || needsStepFree(profile)) {
    return 'Step-free route. Stairs avoided; ramps and elevators used where needed.';
  }
  if (routeType === 'scenic') {
    return 'Gentle route favoring step-free, pleasant walkways.';
  }
  return undefined;
}

export function calculateRoute(
  fromId: string,
  toId: string,
  routeType: RouteType,
  profile: AccessibilityProfile,
  options?: { badWeather?: boolean }
): RouteData | null {
  const fromBuilding = BUILDING_BY_ID.get(fromId);
  const toBuilding = BUILDING_BY_ID.get(toId);
  if (!fromBuilding || !toBuilding) return null;

  const blocked = getBlockedPathIds();
  const badWeather = options?.badWeather ?? isBadWeather();

  const result = dijkstra(fromId, toId, (edge) =>
    edgeCost(edge, routeType, profile, blocked, badWeather)
  );
  if (!result) return null;

  const segments = buildSegments(result.edges);
  const totalDistance = segments.reduce((sum, s) => sum + s.distance_meters, 0);
  const rawWalkTime = segments.reduce((sum, s) => sum + s.walk_time_minutes, 0);
  const totalWalkTime = Math.max(1, Math.round(rawWalkTime * paceMultiplier(profile)));

  const pulseWarnings = collectPulseWarnings(result.path);
  const eventWarnings = blocked.size > 0 ? collectEventWarnings() : undefined;

  return {
    route_type: routeType,
    from_building: fromBuilding,
    to_building: toBuilding,
    steps: buildSteps(result.path, routeType),
    segments,
    waypoints: buildWaypoints(result.path),
    total_distance_meters: totalDistance,
    total_walk_time_minutes: totalWalkTime,
    accessibility_notes: notesForRoute(routeType, profile, badWeather),
    event_warnings: eventWarnings,
    pulse_warnings: pulseWarnings.length > 0 ? pulseWarnings : undefined,
    weather_note: badWeather && routeType === 'weather_shielded'
      ? 'Rain or heat detected — this route stays under cover where possible.'
      : badWeather
      ? 'Consider the Weather-Shielded tab for covered paths.'
      : undefined,
  };
}

export interface RouteOptions {
  fastest?: RouteData;
  accessible?: RouteData;
  scenic?: RouteData;
  quiet?: RouteData;
  weather_shielded?: RouteData;
}

export function calculateRouteOptions(
  fromId: string,
  toId: string,
  profile: AccessibilityProfile,
  options?: { badWeather?: boolean }
): RouteOptions {
  const badWeather = options?.badWeather ?? isBadWeather();
  return {
    fastest: calculateRoute(fromId, toId, 'fastest', profile, { badWeather }) ?? undefined,
    accessible: calculateRoute(fromId, toId, 'accessible', profile, { badWeather }) ?? undefined,
    scenic: calculateRoute(fromId, toId, 'scenic', profile, { badWeather }) ?? undefined,
    quiet: calculateRoute(fromId, toId, 'quiet', profile, { badWeather }) ?? undefined,
    weather_shielded:
      calculateRoute(fromId, toId, 'weather_shielded', profile, { badWeather }) ?? undefined,
  };
}

/** Pick the best default route tab for a user's profile. */
export function primaryRouteKey(profile: AccessibilityProfile, badWeather?: boolean): keyof RouteOptions {
  if (badWeather ?? isBadWeather()) return 'weather_shielded';
  if (profile.sensory_friendly) return 'quiet';
  if (profile.wheelchair || profile.avoid_stairs || profile.elevator_required) return 'accessible';
  return 'fastest';
}

export function getRouteFromOptions(
  options: RouteOptions,
  profile: AccessibilityProfile,
  badWeather?: boolean
): RouteData | undefined {
  const key = primaryRouteKey(profile, badWeather);
  return options[key] ?? options.fastest ?? options.accessible;
}

File: backend/app/routing.py
"""Graph-based campus routing with accessibility-aware edge costs.

Mirrors the client-side engine in services/routing.ts but uses NetworkX, as
described in the CampusIQ architecture (NetworkX route engine).
"""
from __future__ import annotations

import heapq
from typing import Any, Literal

try:  # NetworkX is the primary engine; a pure-Python fallback keeps it runnable anywhere.
    import networkx as nx

    HAS_NETWORKX = True
except ImportError:  # pragma: no cover
    nx = None  # type: ignore
    HAS_NETWORKX = False

from .data import active_events, all_paths, campus_pulse, get_building, get_path_attr, safety_pois

RouteType = Literal["fastest", "accessible", "scenic", "quiet", "weather_shielded"]

BASE_WALK_SPEED = 80  # meters / minute


def _path_id(a: str, b: str) -> str:
    return f"{a}-{b}"


# Adjacency list used for the pure-Python fallback and for edge-data lookups.
def _build_adjacency() -> dict[str, list[dict[str, Any]]]:
    adj: dict[str, list[dict[str, Any]]] = {}

    def add(frm: str, to: str, p: dict[str, Any]) -> None:
        adj.setdefault(frm, []).append(
            {
                "to": to,
                "path_id": _path_id(frm, to),
                "distance": p["distance"],
                "walk_time": p["walk_time"],
                "is_accessible": p["is_accessible"],
                "has_stairs": p["has_stairs"],
            }
        )

    for p in all_paths():
        attr = get_path_attr(p["from"], p["to"])
        extra = {
            "noise_level": attr.get("noise_level", 3),
            "is_covered": attr.get("is_covered", False),
            "is_indoor": attr.get("is_indoor", False),
        }
        add(p["from"], p["to"], {**p, **extra})
        add(p["to"], p["from"], {**p, **extra})
    return adj


_ADJ = _build_adjacency()


def build_graph():
    """Bidirectional directed graph so traversal works both ways."""
    if not HAS_NETWORKX:
        return None
    g = nx.DiGraph()
    for node, edges in _ADJ.items():
        for e in edges:
            g.add_edge(
                node,
                e["to"],
                path_id=e["path_id"],
                distance=e["distance"],
                walk_time=e["walk_time"],
                is_accessible=e["is_accessible"],
                has_stairs=e["has_stairs"],
            )
    return g


_GRAPH = build_graph()


def _dijkstra_fallback(from_id, to_id, weight):
    """Pure-Python Dijkstra used when NetworkX is unavailable."""
    dist = {from_id: 0.0}
    prev: dict[str, str] = {}
    pq: list[tuple[float, str]] = [(0.0, from_id)]
    visited: set[str] = set()

    while pq:
        d, node = heapq.heappop(pq)
        if node in visited:
            continue
        visited.add(node)
        if node == to_id:
            break
        for e in _ADJ.get(node, []):
            cost = weight(node, e["to"], e)
            if cost is None:
                continue
            nd = d + cost
            if nd < dist.get(e["to"], float("inf")):
                dist[e["to"]] = nd
                prev[e["to"]] = node
                heapq.heappush(pq, (nd, e["to"]))

    if to_id not in dist:
        return None
    path = [to_id]
    while path[-1] != from_id:
        if path[-1] not in prev:
            return None
        path.append(prev[path[-1]])
    path.reverse()
    return path


def _edge_data(u: str, v: str) -> dict[str, Any]:
    for e in _ADJ.get(u, []):
        if e["to"] == v:
            return e
    return {}


def blocked_path_ids() -> set[str]:
    blocked: set[str] = set()
    for event in active_events():
        if not event.get("affects_navigation"):
            continue
        for bp in event.get("blocked_paths", []) or []:
            blocked.add(bp)
            parts = bp.split("-")
            if len(parts) == 2:
                blocked.add(_path_id(parts[0], parts[1]))
                blocked.add(_path_id(parts[1], parts[0]))
    return blocked


def _needs_step_free(profile: dict[str, bool]) -> bool:
    return bool(
        profile.get("wheelchair")
        or profile.get("elevator_required")
        or profile.get("avoid_stairs")
    )


def _edge_cost(
    data: dict[str, Any],
    route_type: RouteType,
    profile: dict[str, bool],
    blocked: set[str],
) -> float:
    if data["path_id"] in blocked:
        return float("inf")

    must_be_step_free = route_type in ("accessible", "quiet", "weather_shielded") or _needs_step_free(profile)
    if must_be_step_free and (data["has_stairs"] or not data["is_accessible"]):
        return float("inf")

    cost = float(data["distance"])
    if route_type == "scenic":
        if data["has_stairs"]:
            cost += 250
        if not data["is_accessible"]:
            cost += 150
    elif route_type == "quiet" or profile.get("sensory_friendly"):
        cost += data.get("noise_level", 3) * 35
    elif route_type == "weather_shielded":
        if not data.get("is_covered") and not data.get("is_indoor"):
            cost += 180
    return cost


def _pace_multiplier(profile: dict[str, bool]) -> float:
    if profile.get("slow_walker"):
        return 1.4
    if profile.get("wheelchair"):
        return 1.15
    return 1.0


def calculate_route(
    from_id: str,
    to_id: str,
    route_type: RouteType,
    profile: dict[str, bool],
) -> dict[str, Any] | None:
    from_building = get_building(from_id)
    to_building = get_building(to_id)
    if not from_building or not to_building:
        return None
    if from_id not in _ADJ:
        return None

    blocked = blocked_path_ids()

    def weight(u: str, v: str, data: dict[str, Any]) -> float | None:
        cost = _edge_cost(data, route_type, profile, blocked)
        return None if cost == float("inf") else cost

    if HAS_NETWORKX:
        try:
            node_path = nx.dijkstra_path(_GRAPH, from_id, to_id, weight=weight)
        except (nx.NetworkXNoPath, nx.NodeNotFound):
            return None
    else:
        node_path = _dijkstra_fallback(from_id, to_id, weight)
        if node_path is None:
            return None

    segments: list[dict[str, Any]] = []
    total_distance = 0
    raw_walk_time = 0
    for u, v in zip(node_path, node_path[1:]):
        d = _edge_data(u, v)
        total_distance += d["distance"]
        raw_walk_time += d["walk_time"]
        segments.append(
            {
                "from": u,
                "to": v,
                "distance_meters": d["distance"],
                "walk_time_minutes": d["walk_time"],
                "path_type": "footpath" if d["has_stairs"] else "main_road",
                "is_accessible": d["is_accessible"] and not d["has_stairs"],
                "notes": "Contains stairs" if d["has_stairs"] else None,
            }
        )

    total_walk_time = max(1, round(raw_walk_time * _pace_multiplier(profile)))
    waypoints = [get_building(n)["coordinates"] for n in node_path if get_building(n)]

    steps = _build_steps(node_path, route_type)
    warnings = _event_warnings() if blocked else None

    return {
        "route_type": route_type,
        "from_building": from_building,
        "to_building": to_building,
        "steps": steps,
        "segments": segments,
        "waypoints": waypoints,
        "total_distance_meters": total_distance,
        "total_walk_time_minutes": total_walk_time,
        "accessibility_notes": _accessibility_notes(route_type, profile),
        "event_warnings": warnings,
    }


def calculate_route_options(
    from_id: str, to_id: str, profile: dict[str, bool]
) -> dict[str, Any]:
    return {
        "fastest": calculate_route(from_id, to_id, "fastest", profile),
        "accessible": calculate_route(from_id, to_id, "accessible", profile),
        "scenic": calculate_route(from_id, to_id, "scenic", profile),
        "quiet": calculate_route(from_id, to_id, "quiet", profile),
        "weather_shielded": calculate_route(from_id, to_id, "weather_shielded", profile),
    }


def _build_steps(node_path: list[str], route_type: RouteType) -> list[str]:
    if not node_path:
        return []
    start = get_building(node_path[0])
    steps = [f"Start at {start['name'] if start else node_path[0]}"]
    for i in range(1, len(node_path)):
        b = get_building(node_path[i])
        is_last = i == len(node_path) - 1
        if is_last:
            full = b["name"] if b else node_path[i]
            if route_type == "accessible":
                steps.append(f"Arrive at {full} via the accessible entrance")
            else:
                steps.append(f"Arrive at {full}")
        else:
            name = b["short_name"] if b else node_path[i]
            steps.append(f"Continue past {name}")
    return steps


def _accessibility_notes(route_type: RouteType, profile: dict[str, bool]) -> str | None:
    if route_type == "quiet" or profile.get("sensory_friendly"):
        return "Sensory-quiet route. Avoids noisy cafeterias, crowded plazas, and high-traffic zones."
    if route_type == "weather_shielded":
        return "Weather-shielded route. Maximizes covered walkways and indoor corridors."
    if route_type == "accessible" or _needs_step_free(profile):
        return "Step-free route. Stairs avoided; ramps and elevators used where needed."
    if route_type == "scenic":
        return "Gentle route favoring step-free, pleasant walkways."
    return None


def _event_warnings() -> list[str]:
    return [
        e["alternate_route_note"]
        for e in active_events()
        if e.get("affects_navigation") and e.get("alternate_route_note")
    ]
